### DISCUSSION TITLE
[Feat]: Add Supabase as S3 provider for Coolify

### DISCUSSION THREAD

**Discussion by ukashazia** [coollabsio/coolify#2932] [Feature Requests]
## Objectives:
- create a s3 bucket in supabase
- navigate to `/storages` route in coolify dashboard
- add supabase s3 bucket details (bucket name, region, endpoint, secret key, access key)
- submit

**Comment by luckycreationsindia** [comment]
Isn't it working?

I tried using R2 Storage using S3 by changing endpoint when adding storage details and it's working without any issues.

If I'm not wrong supabase is compatible with S3 Client. So it should work.

Also when I looked into supabase docs I found that it's not fully implemented but some commonly used endpoints will work.

Reference:
https://supabase.com/docs/guides/storage/s3/compatibility

**Comment by ukashazia** [comment]
> Isn't it working?
> 
> I tried using R2 Storage using S3 by changing endpoint when adding storage details and it's working without any issues.
> 
> If I'm not wrong supabase is compatible with S3 Client. So it should work.
> 
> Also when I looked into supabase docs I found that it's not fully implemented but some commonly used endpoints will work.
> 
> Reference: https://supabase.com/docs/guides/storage/s3/compatibility

According to [Coolify's S3 docs](https://coolify.io/docs/knowledge-base/s3-storages), R2 is supported but in the list of compatible S3 providers, there's no supabase S3. I tried linking Supabase's S3 bucket but there seems to be an issue with how the URI is built from the S3 details provided to coolify.

**Comment by ukashazia** [comment]
Screenshot for reference:

![Screenshot_20240502_145414](https://github.com/coollabsio/coolify/assets/92175640/79bf64fd-f6d9-4772-b3f0-b241073a4782)

There seems to be a duplication in url string which should be the reason for the 404.

**Comment by luckycreationsindia** [comment]
Can you share endpoint URL (hiding your project reference from URL). Maybe the Issue is with the endpoint you are providing.

As per the doc it should work.

I just checked that my region is auto (In coolify S3 Settings) when connecting with R2. Can you also try that in supabase endpoint.

And yes it's duplicating the URL string. Make sure you add endpoint as https://[project-ref].supabase.co

This should work.

**Comment by ukashazia** [comment]
> Can you share endpoint URL (hiding your project reference from URL). Maybe the Issue is with the endpoint you are providing.
> 
> As per the doc it should work.
> 
> I just checked that my region is auto (In coolify S3 Settings) when connecting with R2. Can you also try that in supabase endpoint.
> 
> And yes it's duplicating the URL string. Make sure you add endpoint as https://[project-ref].supabase.co
> 
> This should work.

Removing `/storage/v1/s3` from endpoint completely removes it from the resultant URI which throws 404 again.

**Comment by luckycreationsindia** [comment]
Can you share screenshot of error after removing `/storage/v1/s3` from endpoint

Here's my setting for R2 (Can't say about supabase)

Endpoint - https://[account-id].r2.cloudflarestorage.com
Bucket - xxx
Region - auto

**Comment by ukashazia** [comment]
> Can you share screenshot of error after removing `/storage/v1/s3` from endpoint
> 
> Here's my setting for R2 (Can't say about supabase)
> 
> Endpoint - https://[account-id].r2.cloudflarestorage.com Bucket - xxx Region - auto

Sorry if i didn't mention that earlier. I have my R2 setup successful but that's not the case with supabase.

**Comment by luckycreationsindia** [comment]
I tried supabase. There's actually a bug in coolify which is appending anything after slash two times. Maybe this issue is with underlying S3 Client library in coolify.

**Comment by ukashazia** [comment]
> I tried supabase. There's actually a bug in coolify which is appending anything after slash two times. Maybe this issue is with underlying S3 Client library in coolify.

Anyways, i find R2 much better for my current usecase

**Comment by luckycreationsindia** [comment]
> > I tried supabase. There's actually a bug in coolify which is appending anything after slash two times. Maybe this issue is with underlying S3 Client library in coolify.
> 
> Anyways, i find R2 much better for my current usecase

Ok. But still let keep this issue open if anyone wants to use supabase or any other service where endpoint has path included in URL

**Comment by ukashazia** [comment]
> > > I tried supabase. There's actually a bug in coolify which is appending anything after slash two times. Maybe this issue is with underlying S3 Client library in coolify.
> > 
> > 
> > Anyways, i find R2 much better for my current usecase
> 
> Ok. But still let keep this issue open if anyone wants to use supabase or any other service where endpoint has path included in URL

i might as well contributed and fixed it it was written in js. looks like it's implemented in PHP :)

**Comment by luckycreationsindia** [comment]
> > > > I tried supabase. There's actually a bug in coolify which is appending anything after slash two times. Maybe this issue is with underlying S3 Client library in coolify.
> > > 
> > > 
> > > Anyways, i find R2 much better for my current usecase
> > 
> > 
> > Ok. But still let keep this issue open if anyone wants to use supabase or any other service where endpoint has path included in URL
> 
> i might as well contributed and fixed it it was written in js. looks like it's implemented in PHP :)

Same here. I tried checking the code it's using "flysystem-aws-s3-v3" library. Checked the code where it's concatenating the endpoint with bucket. Couldn't find more as I don't know how PHP with Laravel actually works. I just assumed it's issue with the library.

**Comment by actraiser** [comment]
Out of curiosity, how do you configure a Supabase S3 Storage in the UI? I use R2 with Supabase but what I did is to hardcode the R2-Endpoint, Bucket and Region in the docker-compose.yml under the supabase-storage-service and changing Minio User/Password in the Coolify UI to reflect the credentials on R2.

```
    environment:
      - STORAGE_BACKEND=s3
      - STORAGE_S3_BUCKET=[R2 bucket]
      - STORAGE_S3_ENDPOINT=[R2 endpoint] # replaces http://supabase-minio:9000
      - STORAGE_S3_REGION=[R2 region]
```

Is there a simpler way using the Coolify or Supabase UI directly?

Greets
-act

**Comment by ukashazia** [comment]
> Out of curiosity, how do you configure a Supabase S3 Storage in the UI?

You can add an S3 bucket in `/storages`. But this bucket can only be used by coolify to store backups of databases as far as i know.
After you've successfully added a bucket, you have to select it from `backups` section/screen of a database to backup your db backups to S3.
.

**Comment by actraiser** [comment]
> You can add an S3 bucket in `/storages`. But this bucket can only be used by coolify to store backups of databases as far as i know.

Ah sorry, i misunderstood - you are talking about database backups to S3 and not about using an external S3 Bucket for the Supabase Storage API (to have an external CDN). You can actually easily use the Coolify configured S3 Storage for Supabase database backups by going to your Supabase resource, then scroll all the way down to the supabase-db service in the list and click on settings. there is a "backups" option that you can configure to schedule backups to your S3 buckets you have configured with Coolify before. At least for me - R2 works totally fine.

<img width="1066" alt="Bildschirmfoto 2024-05-04 um 08 17 45" src="https://github.com/coollabsio/coolify/assets/5455/1196c5e4-b654-4c15-884c-5d447b8e6d72">

And it will show up in R2:

<img width="1089" alt="Bildschirmfoto 2024-05-04 um 08 21 22" src="https://github.com/coollabsio/coolify/assets/5455/bd94a579-996e-4ad5-9078-e968efb35eaf">


Greets
-act

**Comment by ukashazia** [comment]
We were trying to add S3 as a db backup location to coolify. R2 works but supabase's hosted/managed S3 (not the selfhosted one) is having some issues.

**Comment by luckycreationsindia** [comment]
As per document I think supabase is not tested and is not supported.

https://coolify.io/docs/knowledge-base/s3-storages

**Comment by andrasbacsai** [comment]
Thank you for the suggestions. Features should be added to the `Discussion` section, so I will convert this one to it.

**Reply by jadams777** [reply]
So how do I connect my S3 storage bucket to SupaBase in Coolify UI? Not for backups, but for object storage.

**Comment by aragornii** [comment]
Hi all, any news on this issue? I'm facing the same behavior. This is the error I'm getting:

```
Unable to list contents for '', shallow listing

Reason: Error executing "ListObjectsV2" on "https://[my-project].supabase.co/storage/v1/s3/storage/v1/s3/backup-coolify-db?list-type=2&prefix=&delimiter=%2F";
AWS HTTP error: Client error: `GET https://[my-project].supabase.co/storage/v1/s3/storage/v1/s3/backup-coolify-db?list-type=2&prefix=&delimiter=%2F` resulted in a `404 Not Found` response:
storage/v1/s3/backup-coolify-db (truncated...)
 NoSuchKey (client): Object not found - storage/v1/s3/backup-coolify-dbNoSuchKeyObject not found
```

As mentioned above, I think it depends on the fact coolify is appending `storage/v1/s3/` twice. However, if I try to remove it I get:

```
Reason: Error executing "ListObjectsV2" on "https://[my-project].supabase.co/backup-coolify-db?list-type=2&prefix=&delimiter=%2F";
AWS HTTP error: Client error: `GET https://[my-project].supabase.co/backup-coolify-db?list-type=2&prefix=&delimiter=%2F` resulted in a `404 Not Found` response:
{"error":"requested path is invalid"}
 Unable to parse error information from response - Error parsing XML: String could not be parsed as XML
```

Thank you.

**Comment by robinsadeghpour** [comment]
+1 need that to

**Comment by hassankhan2608** [comment]
is this fixed ???

**Comment by andrasbacsai** [comment]
yes, this works.
